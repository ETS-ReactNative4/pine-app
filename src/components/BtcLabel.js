import React, { Component } from 'react';
import PropTypes from 'prop-types';
import StyledText from './StyledText';

const UNIT_BTC = 'BTC';
const UNIT_MBTC = 'mBTC';
const UNIT_SATOSHIS = 'Satoshis';

const splitNumber = (number) => {
  const parts = number.split('.');
  const integer = parts[0];
  const fractional = parts[1];

  return { integer, fractional };
};

const trimTrailingZeros = (number) => {
  if (!number) {
    return '';
  }

  return number.replace(/[0]+$/, '');
};

const getUnitLabel = (unit) => {
  if (unit === UNIT_SATOSHIS) {
    return 'sats';
  }

  return unit;
};

const reverseString = (string) => {
  return string.split('').reverse().join('');
};

const addThousandsSeparators = (integer) => {
  if (!integer || integer.length < 4) {
    return integer;
  }

  const reversedInteger = reverseString(integer);
  const maskedReversedInteger = reversedInteger.match(/.{1,3}/g).join(' ');
  const maskedInteger = reverseString(maskedReversedInteger);

  return maskedInteger;
};

const formatBtc = (amount) => {
  const amountString = amount.toFixed(8);
  const number = splitNumber(amountString);

  number.fractional = trimTrailingZeros(number.fractional);

  return number;
};

const formatMBtc = (amount) => {
  const amountString = (amount * 1000).toFixed(11);
  const number = splitNumber(amountString);

  number.fractional = trimTrailingZeros(number.fractional);

  return number;
};

const formatSatoshis = (amount) => {
  let satoshis = amount * 100000000;
  let amountString;
  let suffix = '';

  if (satoshis > 1000000) {
    satoshis /= 1000000;
    suffix = 'M';
    amountString = satoshis.toFixed(6);
  } else if (satoshis > 1000) {
    satoshis /= 1000;
    suffix = 'k';
    amountString = satoshis.toFixed(3);
  } else {
    amountString = satoshis.toFixed(0);
  }

  const number = splitNumber(amountString);

  number.fractional = trimTrailingZeros(number.fractional);
  number.suffix = suffix;

  return number;
};

const formatAmount = (amount, unit, hideUnit) => {
  const unitLabel = hideUnit ? '' : getUnitLabel(unit);

  if (!amount) {
    return `0 ${unitLabel}`.trim();
  }

  let number;

  switch (unit) {
    case UNIT_BTC:
      number = formatBtc(amount);
      break;

    case UNIT_MBTC:
      number = formatMBtc(amount);
      break;

    case UNIT_SATOSHIS:
      number = formatSatoshis(amount);
      break;

    default:
      return '‚ùì';
  }

  number.integer = addThousandsSeparators(number.integer);
  number.suffix = number.suffix || '';

  if (!number.fractional) {
    return `${number.integer}${number.suffix} ${unitLabel}`.trim();
  }

  return `${number.integer}.${number.fractional}${number.suffix} ${unitLabel}`.trim();
};

export default class BtcLabel extends Component {
  render() {
    const { amount, unit, hideUnit } = this.props;
    const label = formatAmount(amount, unit, hideUnit);

    return (
      <StyledText style={this.props.style}>
        {label}
      </StyledText>
    );
  }
}

BtcLabel.propTypes = {
  style: PropTypes.any,
  amount: PropTypes.number.isRequired,
  unit: PropTypes.string,
  hideUnit: PropTypes.bool
};
