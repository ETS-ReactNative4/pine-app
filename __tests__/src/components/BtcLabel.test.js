import React from 'react';
import renderer from 'react-test-renderer';
import BtcLabel from '../../../src/components/BtcLabel';

const UNIT_BTC = 'BTC';
const UNIT_MBTC = 'mBTC';
const UNIT_SATOSHIS = 'Satoshis';

describe('BtcLabel', () => {
  describe('when unit is BTC', () => {
    it('renders 0 BTC correctly', () => {
      const tree = renderer.create(
        <BtcLabel amount={0} unit={UNIT_BTC} />
      ).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('renders 2 BTC correctly', () => {
      const tree = renderer.create(
        <BtcLabel amount={2} unit={UNIT_BTC} />
      ).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('renders 2.00 BTC correctly', () => {
      const tree = renderer.create(
        <BtcLabel amount={2.00} unit={UNIT_BTC} />
      ).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('renders 2.30 BTC correctly', () => {
      const tree = renderer.create(
        <BtcLabel amount={2.30} unit={UNIT_BTC} />
      ).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('renders 0.00002001 BTC correctly', () => {
      const tree = renderer.create(
        <BtcLabel amount={0.00002001} unit={UNIT_BTC} />
      ).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('renders 0.00000002 BTC correctly', () => {
      const tree = renderer.create(
        <BtcLabel amount={0.00000002} unit={UNIT_BTC} />
      ).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('renders 4000 BTC correctly', () => {
      const tree = renderer.create(
        <BtcLabel amount={4000} unit={UNIT_BTC} />
      ).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });

  describe('when unit is mBTC', () => {
    it('renders 0 BTC correctly', () => {
      const tree = renderer.create(
        <BtcLabel amount={0} unit={UNIT_MBTC} />
      ).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('renders 2 BTC correctly', () => {
      const tree = renderer.create(
        <BtcLabel amount={2} unit={UNIT_MBTC} />
      ).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('renders 0.00002001 BTC correctly', () => {
      const tree = renderer.create(
        <BtcLabel amount={0.00002001} unit={UNIT_MBTC} />
      ).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('renders 0.00000002 BTC correctly', () => {
      const tree = renderer.create(
        <BtcLabel amount={0.00000002} unit={UNIT_MBTC} />
      ).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });

  describe('when unit is Satoshis', () => {
    it('renders 0 BTC correctly', () => {
      const tree = renderer.create(
        <BtcLabel amount={0} unit={UNIT_SATOSHIS} />
      ).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('renders 2 BTC correctly', () => {
      const tree = renderer.create(
        <BtcLabel amount={2} unit={UNIT_SATOSHIS} />
      ).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('renders 0.00002001 BTC correctly', () => {
      const tree = renderer.create(
        <BtcLabel amount={0.00002001} unit={UNIT_SATOSHIS} />
      ).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('renders 0.00000002 BTC correctly', () => {
      const tree = renderer.create(
        <BtcLabel amount={0.00000002} unit={UNIT_SATOSHIS} />
      ).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });

  describe('when unit is undefined', () => {
    it('renders without errors', () => {
      const tree = renderer.create(
        <BtcLabel amount={123} />
      ).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });
});
