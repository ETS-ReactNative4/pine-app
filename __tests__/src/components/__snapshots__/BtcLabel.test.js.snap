// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BtcLabel when unit is BTC renders -0.02784 BTC correctly 1`] = `
<Text
  style={
    Array [
      Object {
        "fontFamily": "System",
        "fontWeight": "400",
      },
      Object {
        "color": "#000000",
      },
      undefined,
    ]
  }
>
  -
  2,784,000 sats
</Text>
`;

exports[`BtcLabel when unit is BTC renders 0 BTC correctly 1`] = `
<Text
  style={
    Array [
      Object {
        "fontFamily": "System",
        "fontWeight": "400",
      },
      Object {
        "color": "#000000",
      },
      undefined,
    ]
  }
>
  0 BTC
</Text>
`;

exports[`BtcLabel when unit is BTC renders 0.00000002 BTC correctly 1`] = `
<Text
  style={
    Array [
      Object {
        "fontFamily": "System",
        "fontWeight": "400",
      },
      Object {
        "color": "#000000",
      },
      undefined,
    ]
  }
>
  0.00000002 BTC
</Text>
`;

exports[`BtcLabel when unit is BTC renders 0.00002001 BTC correctly 1`] = `
<Text
  style={
    Array [
      Object {
        "fontFamily": "System",
        "fontWeight": "400",
      },
      Object {
        "color": "#000000",
      },
      undefined,
    ]
  }
>
  0.00002001 BTC
</Text>
`;

exports[`BtcLabel when unit is BTC renders 2 BTC correctly 1`] = `
<Text
  style={
    Array [
      Object {
        "fontFamily": "System",
        "fontWeight": "400",
      },
      Object {
        "color": "#000000",
      },
      undefined,
    ]
  }
>
  2 BTC
</Text>
`;

exports[`BtcLabel when unit is BTC renders 2.00 BTC correctly 1`] = `
<Text
  style={
    Array [
      Object {
        "fontFamily": "System",
        "fontWeight": "400",
      },
      Object {
        "color": "#000000",
      },
      undefined,
    ]
  }
>
  2 BTC
</Text>
`;

exports[`BtcLabel when unit is BTC renders 2.30 BTC correctly 1`] = `
<Text
  style={
    Array [
      Object {
        "fontFamily": "System",
        "fontWeight": "400",
      },
      Object {
        "color": "#000000",
      },
      undefined,
    ]
  }
>
  2.3 BTC
</Text>
`;

exports[`BtcLabel when unit is BTC renders 4000 BTC correctly 1`] = `
<Text
  style={
    Array [
      Object {
        "fontFamily": "System",
        "fontWeight": "400",
      },
      Object {
        "color": "#000000",
      },
      undefined,
    ]
  }
>
  4,000 BTC
</Text>
`;

exports[`BtcLabel when unit is Satoshis renders -0.08 BTC correctly 1`] = `
<Text
  style={
    Array [
      Object {
        "fontFamily": "System",
        "fontWeight": "400",
      },
      Object {
        "color": "#000000",
      },
      undefined,
    ]
  }
>
  -
  8,000,000 sats
</Text>
`;

exports[`BtcLabel when unit is Satoshis renders 0 BTC correctly 1`] = `
<Text
  style={
    Array [
      Object {
        "fontFamily": "System",
        "fontWeight": "400",
      },
      Object {
        "color": "#000000",
      },
      undefined,
    ]
  }
>
  0 sats
</Text>
`;

exports[`BtcLabel when unit is Satoshis renders 0.00000002 BTC correctly 1`] = `
<Text
  style={
    Array [
      Object {
        "fontFamily": "System",
        "fontWeight": "400",
      },
      Object {
        "color": "#000000",
      },
      undefined,
    ]
  }
>
  2 sats
</Text>
`;

exports[`BtcLabel when unit is Satoshis renders 0.00002001 BTC correctly 1`] = `
<Text
  style={
    Array [
      Object {
        "fontFamily": "System",
        "fontWeight": "400",
      },
      Object {
        "color": "#000000",
      },
      undefined,
    ]
  }
>
  2,001 sats
</Text>
`;

exports[`BtcLabel when unit is Satoshis renders 2 BTC correctly 1`] = `
<Text
  style={
    Array [
      Object {
        "fontFamily": "System",
        "fontWeight": "400",
      },
      Object {
        "color": "#000000",
      },
      undefined,
    ]
  }
>
  200,000,000 sats
</Text>
`;

exports[`BtcLabel when unit is mBTC renders -1.4234 BTC correctly 1`] = `
<Text
  style={
    Array [
      Object {
        "fontFamily": "System",
        "fontWeight": "400",
      },
      Object {
        "color": "#000000",
      },
      undefined,
    ]
  }
>
  -
  1,423.4 mBTC
</Text>
`;

exports[`BtcLabel when unit is mBTC renders 0 BTC correctly 1`] = `
<Text
  style={
    Array [
      Object {
        "fontFamily": "System",
        "fontWeight": "400",
      },
      Object {
        "color": "#000000",
      },
      undefined,
    ]
  }
>
  0 mBTC
</Text>
`;

exports[`BtcLabel when unit is mBTC renders 0.00000002 BTC correctly 1`] = `
<Text
  style={
    Array [
      Object {
        "fontFamily": "System",
        "fontWeight": "400",
      },
      Object {
        "color": "#000000",
      },
      undefined,
    ]
  }
>
  0.00002 mBTC
</Text>
`;

exports[`BtcLabel when unit is mBTC renders 0.00002001 BTC correctly 1`] = `
<Text
  style={
    Array [
      Object {
        "fontFamily": "System",
        "fontWeight": "400",
      },
      Object {
        "color": "#000000",
      },
      undefined,
    ]
  }
>
  0.02001 mBTC
</Text>
`;

exports[`BtcLabel when unit is mBTC renders 2 BTC correctly 1`] = `
<Text
  style={
    Array [
      Object {
        "fontFamily": "System",
        "fontWeight": "400",
      },
      Object {
        "color": "#000000",
      },
      undefined,
    ]
  }
>
  2,000 mBTC
</Text>
`;

exports[`BtcLabel when unit is undefined renders without errors 1`] = `
<Text
  style={
    Array [
      Object {
        "fontFamily": "System",
        "fontWeight": "400",
      },
      Object {
        "color": "#000000",
      },
      undefined,
    ]
  }
>
  ‚ùì
</Text>
`;
